truenas_verify/mtree_verify.py

```diff
@@ -0,0 +1,128 @@
+from collections import namedtuple
+from hashlib import file_digest
+from multiprocessing import Pool
+from os import lstat
+from stat import S_ISDIR, S_ISREG, S_ISLNK, S_IMODE
+import sys
+
+
+LOG_PATH = '/var/log/truenas_verify.log'
+MTREE_FILE_PATH = '/conf/rootfs.mtree'
+MTREE_ENTRY = namedtuple('MtreeEntry', ['fname', 'mode', 'uid', 'gid', 'type', 'link', 'size', 'sha256'])
+
+
+def parse_mtree_entry(line):
+    if line.startswith('#'):
+        return None
+
+    fname, mode, gid, uid, extra = line[1:].split(maxsplit=4)
+    if extra.startswith('type=dir'):
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            'dir',
+            None,
+            None,
+            None
+        )
+    elif extra.startswith('type=link'):
+        ftype, link = extra.split()
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            'link',
+            link.split('=')[1],
+            None,
+            None
+        )
+    else:
+        ftype, size, shasum = extra.split()
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            ftype.split('=')[1],
+            None,
+            int(size.split('=')[1]),
+            shasum.split('=')[1]
+        )
+
+    return entry
+
+
+def validate_file_sha256sum(entry, errors):
+    with open(entry.fname, 'rb', buffering=0) as f:
+        hash = file_digest(f, 'sha256').hexdigest()
+        if hash != entry.sha256:
+            errors.append(f'{entry.fname}: expected: {entry.sha256}, got: {hash}')
+
+
+def validate_mtree_entry(entry) -> list[str]:
+    try:
+        st = lstat(entry.fname)
+    except FileNotFoundError:
+        return [f'{entry.fname}: file does not exist.']
+
+    assert st.st_uid == entry.uid, entry.fname
+    assert st.st_gid == entry.gid, entry.fname
+
+    errors = []
+    match entry.type:
+        case 'dir':
+            if not S_ISDIR(st.st_mode):
+                errors.append(f'{entry.fname}: incorrect file type.')
+        case 'file':
+            if not S_ISREG(st.st_mode):
+                errors.append(f'{entry.fname}: incorrect file type.')
+
+            validate_file_sha256sum(entry, errors)
+        case 'link':
+            if not S_ISLNK(st.st_mode):
+                errors.append(f'{entry.fname}: incorrect file type.')
+
+    if oct(S_IMODE(st.st_mode))[2:] != entry.mode:
+        errors.append(f'{entry.fname}: got mode {oct(S_IMODE(st.st_mode))}, expected: {entry.mode}')
+
+    return errors
+
+
+def load_mtree_file():
+    with open(MTREE_FILE_PATH, 'r') as f:
+        chunk_size = 1000
+        chunk = []
+        for line in f:
```

Comment:
We could probably use this: https://github.com/truenas/middleware/blob/604a5da824c2fefcfcf4ce1669b7fd057ff6bc01/src/middlewared/middlewared/utils/itertools.py#L29

For batching up contents. With intention to use `itertools.batched` in python 3.12.

Reply:
I'm unsure about making this package depend on middlewared when middlewared already depends on it. But I see some of our other repos doing it so I guess it's ok?

Reply:
Well, I meant you could probably copy-paste method (since it's pretty tiny) with the comment with goal of replacing when there's cpython-optimized batched in python3 3.12.